### PR TITLE
Fix pipeline set up with multi-stream capture

### PR DESCRIPTION
**Pull request by ranj063**

### CODE REVIEW

File: sound/soc/sof/sof-audio.c [hunk 1]
```diff
@@ -342,27 +349,31 @@ sof_prepare_widgets_in_path(struct snd_sof_dev *sdev, struct snd_soc_dapm_widget
...
 		if (!p->walking && p->sink->dobj.private && widget->dobj.private) {
+			if (!widget_in_list(list, p->sink))
+				continue;
+
 			p->walking = true;
-			if (WIDGET_IS_AIF_OR_DAI(widget->id)) {
-				err = sof_widget_free(sdev, widget->dobj.private);
-				if (err < 0)
-					ret = err;
-			}
```
Comment (plbossart) [inline sound/soc/sof/sof-audio.c:359]: for my education, why did you move this block to before the loop @ranj063 ? It's not clear to me why we special case

File: sound/soc/sof/sof-audio.c [hunk 2]
```diff
@@ -378,19 +389,23 @@ static int sof_free_widgets_in_path(struct snd_sof_dev *sdev, struct snd_soc_dap
...
 		if (!p->walking && p->sink->dobj.private && widget->dobj.private) {
+			if (!widget_in_list(list, p->sink))
+				continue;
+
 			p->walking = true;
-			if (WIDGET_IS_AIF_OR_DAI(widget->id)) {
-				ret = sof_widget_setup(sdev, widget->dobj.private);
-				if (ret < 0)
-					goto out;
-			}
```
Comment (plbossart) [inline sound/soc/sof/sof-audio.c:393]: same here

File: sound/soc/sof/sof-audio.c (hunk 1 above)
Comment (ranj063) [reply sound/soc/sof/sof-audio.c:359]: @plbossart this was a mistake in the first place. Assuming we have a DAI copier connected to 2 different host copiers, we would set up the DAI copier widget twice, once for each sink path.

Comment (plbossart) [reply sound/soc/sof/sof-audio.c:359]: should this be treated as a separate fix then, and in a second patch add the check on the widget path? It's not clear to be if the two parts are strictly related?

Comment (ranj063) [reply sound/soc/sof/sof-audio.c:359]: yes, fixed now. Thanks!

Comment (plbossart) [reply sound/soc/sof/sof-audio.c:359]: ok, thanks

**Review by plbossart — DISMISSED** [review]

File: sound/soc/sof/sof-audio.c [hunk 3]
```diff
@@ -348,15 +348,16 @@ static int sof_free_widgets_in_path(struct snd_sof_dev *sdev, struct snd_soc_dap
 	int err;
 	int ret = 0;
 
+	if (WIDGET_IS_AIF_OR_DAI(widget->id)) {
```
Comment (bardliao) [inline sound/soc/sof/sof-audio.c]: The existing code test `widget->dobj.private` before calling `sof_widget_free(sdev, widget->dobj.private);`. Should we do the same test when we move `sof_widget_free(sdev, widget->dobj.private);` out of the loop?

File: sound/soc/sof/sof-audio.c [hunk 4]
```diff
@@ -348,15 +348,16 @@ static int sof_free_widgets_in_path(struct snd_sof_dev *sdev, struct snd_soc_dap
...
 	snd_soc_dapm_widget_for_each_sink_path(widget, p) {
 		if (!p->walking && p->sink->dobj.private && widget->dobj.private) {
 			p->walking = true;
-			if (WIDGET_IS_AIF_OR_DAI(widget->id)) {
-				err = sof_widget_free(sdev, widget->dobj.private);
-				if (err < 0)
-					ret = err;
-			}
 
 			err = sof_widget_free(sdev, p->sink->dobj.private);
```
Comment (bardliao) [inline sound/soc/sof/sof-audio.c]: Can we call `sof_widget_free(sdev, widget->dobj.private);` unconditionally outside the loop and remove `sof_widget_free(sdev, p->sink->dobj.private);` here?

File: sound/soc/sof/sof-audio.c (hunk 3 above)
Comment (ranj063) [reply sound/soc/sof/sof-audio.c]: good point. added the check now

File: sound/soc/sof/sof-audio.c (hunk 4 above)
Comment (ranj063) [reply sound/soc/sof/sof-audio.c]: yes, i've fixed both setup and free functions to unconditionally set up the starting widget outside the loop now

**Review by bardliao — APPROVED** [review]

**Review by bardliao — APPROVED** [review]

File: sound/soc/sof/sof-audio.c [hunk 5]
```diff
@@ -383,32 +380,23 @@ static int sof_set_up_widgets_in_path(struct snd_sof_dev *sdev, struct snd_soc_d
...
-				if (WIDGET_IS_AIF_OR_DAI(widget->id))
-					sof_widget_free(sdev, widget->dobj.private);
-				goto out;
-			}
 
 			ret = sof_set_up_widgets_in_path(sdev, p->sink, dir);
+			p->walking = false;
 			if (ret < 0) {
-				if (WIDGET_IS_AIF_OR_DAI(widget->id))
+				if (widget->dobj.private)
```
Comment (plbossart) [inline sound/soc/sof/sof-audio.c:409]: can this condition ever be false?

File: sound/soc/sof/sof-audio.c [hunk 6]
```diff
@@ -282,9 +283,11 @@ sof_unprepare_widgets_in_path(struct snd_sof_dev *sdev, struct snd_soc_dapm_widg
 
 	/* unprepare all widgets in the sink paths */
 	snd_soc_dapm_widget_for_each_sink_path(widget, p) {
+		if (!widget_in_list(list, p->sink))
+			continue;
```
Comment (ujfalusi) [inline sound/soc/sof/sof-audio.c:287]: How this happens?
We walk from source to sink, so walk through the widgets which are in the path.
How we are enabling a widget (which is **in** the path) which is **not in** the path?

File: sound/soc/sof/sof-audio.c [hunk 7]
```diff
@@ -383,32 +380,23 @@ static int sof_set_up_widgets_in_path(struct snd_sof_dev *sdev, struct snd_soc_d
...
-				if (WIDGET_IS_AIF_OR_DAI(widget->id))
+				if (widget->dobj.private)
 					sof_widget_free(sdev, widget->dobj.private);
-				sof_widget_free(sdev, p->sink->dobj.private);
-			}
-out:
-			p->walking = false;
-			if (ret < 0)
 				return ret;
+			}
```
Comment (plbossart) [inline sound/soc/sof/sof-audio.c:412]: question: is this error handling sufficient? Doing a single call to sof_widget_free() in a loop will leave all other widgets setup. That's the traditional issue with error handling inside a loop. Do we need an unwind goto instead?

**Review by plbossart — COMMENTED** [review]
Looks good but two questions on error handling (I haven't changed have I)

File: sound/soc/sof/sof-audio.c (hunk 5 above)
Comment (ranj063) [reply sound/soc/sof/sof-audio.c:409]: yes, the driver doesnt handle all widget types and esp the virtual widgets we have in topology will have this condition false.

File: sound/soc/sof/sof-audio.c (hunk 7 above)
Comment (ranj063) [reply sound/soc/sof/sof-audio.c:412]: yes, I think this is sufficient. If a recusive call to sof_widget_setup() fails, we will free the widget that was set up before the loop in invoked in every recursive call.

File: sound/soc/sof/sof-audio.c (hunk 6 above)
Comment (ranj063) [reply sound/soc/sof/sof-audio.c:287]: @ujfalusi snd_soc_dapm_widget_for_each_sink_path() looks at all paths leaving the widget not just the ones that have an active DAPM widget associated with the PCM.

**Review by plbossart — APPROVED** [review]

**Comment by plbossart** [comment]
@ujfalusi are you good with this PR, @bardliao tells me it's required for the multi-stream capture support so this needs to be merged soonish.

**Review by ujfalusi — APPROVED** [review]

File: sound/soc/sof/sof-audio.c (hunk 6 above)
Comment (ujfalusi) [reply sound/soc/sof/sof-audio.c:287]: @ranj063, oh yes, I see now, thanks!
